File: 14-errors/main.go
package main

import (
	"errors"
	"fmt"
)

// 1. 定义哨兵错误
var (
	ErrNotFound         = errors.New("not found")
	ErrPermissionDenied = errors.New("permission denied")
	ErrDivisionByZero   = errors.New("division by zero")
)

// 2. 自定义错误类型
type FileError struct {
	Path    string
	Message string
}

func (e *FileError) Error() string {
	return fmt.Sprintf("file error: %s: %s", e.Path, e.Message)
}

// 3. (T, error) 模式
func divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, ErrDivisionByZero
	}
	return a / b, nil
}

// 4. 错误包装
func readConfig(path string) (string, error) {
	if path == "missing.txt" {
		return "", fmt.Errorf("readConfig: %w", ErrNotFound)
	}
	if path == "secret.txt" {
		return "", &FileError{Path: path, Message: "permission denied"}
	}
	return "config_data", nil
}

// 5. 错误传播 - Go 的 if err != nil 模式
func loadAndProcess(path string) (string, error) {
	data, err := readConfig(path)
	if err != nil {
		return "", fmt.Errorf("loadAndProcess: %w", err)
	}
	_ = data
	return "processed_data", nil
}

func main() {
	// 1. 基本错误处理
	result, err := divide(10, 3)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("10 / 3 = %.2f\n", result)

	// 2. 处理除零错误
	_, err = divide(10, 0)
	if err != nil {
		fmt.Println("10 / 0 error:", err)
	}

	// 3. errors.Is - 判断错误链中是否包含特定错误
	_, err = loadAndProcess("missing.txt")
	if errors.Is(err, ErrNotFound) {
		fmt.Println("File not found (detected via errors.Is)")
	}

	// 4. errors.As - 提取特定类型的错误
	_, err = readConfig("secret.txt")
	var fileErr *FileError
	if errors.As(err, &fileErr) {
		fmt.Printf("FileError on path %q: %s\n", fileErr.Path, fileErr.Message)
	}

	// 5. 成功路径
	config, err := readConfig("normal.txt")
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Println("Good config:", config)

	// 注意：Go 允许忽略错误返回值，编译器不会警告
	// result2, _ := divide(10, 0)  // 忽略错误，这是合法但危险的
}

File: 14-errors/main.rs
use std::fmt;
use std::num::ParseIntError;

// 1. 自定义错误类型
#[derive(Debug)]
enum AppError {
    NotFound(String),
    PermissionDenied,
    DivisionByZero,
    ParseError(ParseIntError),
}

// 实现 Display trait
impl fmt::Display for AppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AppError::NotFound(path) => write!(f, "not found: {}", path),
            AppError::PermissionDenied => write!(f, "permission denied"),
            AppError::DivisionByZero => write!(f, "division by zero"),
            AppError::ParseError(e) => write!(f, "parse error: {}", e),
        }
    }
}

// 2. From trait 实现自动错误转换
impl From<ParseIntError> for AppError {
    fn from(e: ParseIntError) -> Self {
        AppError::ParseError(e)
    }
}

// 3. Result<T, E> 返回类型
fn divide(a: f64, b: f64) -> Result<f64, AppError> {
    if b == 0.0 {
        return Err(AppError::DivisionByZero);
    }
    Ok(a / b)
}

fn read_config(path: &str) -> Result<String, AppError> {
    if path == "missing.txt" {
        return Err(AppError::NotFound(path.to_string()));
    }
    if path == "secret.txt" {
        return Err(AppError::PermissionDenied);
    }
    Ok("config_data".to_string())
}

// 4. ? 运算符 - 自动传播错误（类似 Zig 的 try）
fn load_and_process(path: &str) -> Result<String, AppError> {
    let _data = read_config(path)?; // 错误时自动返回 Err
    Ok("processed_data".to_string())
}

// 5. ? 运算符 + From trait 自动转换
fn parse_and_double(s: &str) -> Result<i32, AppError> {
    let num: i32 = s.parse()?; // ParseIntError 自动转换为 AppError
    Ok(num * 2)
}

fn main() {
    // 1. 基本错误处理 - match
    match divide(10.0, 3.0) {
        Ok(result) => println!("10 / 3 = {:.2}", result),
        Err(e) => println!("Error: {}", e),
    }

    // 2. 处理除零
    match divide(10.0, 0.0) {
        Ok(result) => println!("Result: {}", result),
        Err(e) => println!("10 / 0 error: {}", e),
    }

    // 3. unwrap_or - 提供默认值（类似 Zig 的 catch）
    let safe_result = divide(10.0, 0.0).unwrap_or(0.0);
    println!("10 / 0 (with default) = {:.2}", safe_result);

    // 4. if let - 只关心成功情况
    if let Ok(config) = read_config("normal.txt") {
        println!("Good config: {}", config);
    }

    // 5. 错误匹配
    match read_config("missing.txt") {
        Ok(data) => println!("Data: {}", data),
        Err(AppError::NotFound(path)) => println!("File {} not found", path),
        Err(AppError::PermissionDenied) => println!("Permission denied!"),
        Err(e) => println!("Other error: {}", e),
    }

    // 6. ? 运算符传播（在 main 中演示需要用 match 包裹）
    match load_and_process("missing.txt") {
        Ok(data) => println!("Processed: {}", data),
        Err(e) => println!("load_and_process failed: {}", e),
    }

    // 7. From trait 自动错误转换
    match parse_and_double("42") {
        Ok(v) => println!("Parsed and doubled: {}", v),
        Err(e) => println!("Error: {}", e),
    }
    match parse_and_double("abc") {
        Ok(v) => println!("Parsed: {}", v),
        Err(e) => println!("Parse error: {}", e),
    }

    // 8. map / and_then 组合子
    let result = divide(10.0, 2.0)
        .map(|v| v as i32)
        .map(|v| v * 3);
    println!("Chained result: {:?}", result); // Ok(15)
}
